onnxruntime/core/optimizer/qdq_transformer/qdq_propagation.cc

```diff
@@ -228,21 +308,48 @@ Status PropagateDQForward(Graph& graph, gsl::span<const NodeIndex> node_indices,
                               ? dq_node.MutableInputDefs()[QDQ::InputIndex::ZERO_POINT_ID]
                               : nullptr;
 
-    const auto edge_after_dq = GetNextEdge(graph, dq_node);
-    if (!edge_after_dq) {
+    const InlinedVector<ExtendedGraphEdge> edges_after_dq = GetNextEdges(graph, dq_node);
+    if (edges_after_dq.size() != 1) {
       continue;
     }
 
-    for (auto curr_edge = GetNextPropagationEdge(graph, *edge_after_dq);
-         curr_edge.has_value();
-         curr_edge = GetNextPropagationEdge(graph, *curr_edge)) {
-      if (const auto* dst_node = curr_edge->GetNodeAtEnd(graph, ExtendedGraphEdge::End::Destination);
-          dst_node && QDQ::MatchQNode(*dst_node)) {
-        break;
+    // Utility function to check if any edge out of a node (e.g., Transpose) ends in a Q node.
+    auto any_edge_ends_in_q = [](Graph& graph, const InlinedVector<ExtendedGraphEdge>& edges) -> bool {
+      for (const auto& edge : edges) {
+        const auto* edge_dst_node = edge.GetNodeAtEnd(graph, ExtendedGraphEdge::End::Destination);
+        if (edge_dst_node && QDQ::MatchQNode(*edge_dst_node)) {
+          return true;
+        }
+      }
+      return false;
+    };
+
+    // Propagate DQ forward in a BFS traversal of "edge groups". A single edge group consists of one or more edges
+    // that all begin at a unique source node and end at one or more destination nodes. Ex: The subgraph below shows
+    // an edge group (containing 3 edges) that begins at a Transpose, ends at two destination nodes, and produces a
+    // graph output.
+    //    DQ -> Transpose --+--> Sigmoid -> ...
+    //                      |
+    //                      +--> Slice -> ...
+    //                      |
+    //                      +--> graph_output
+    std::queue<InlinedVector<ExtendedGraphEdge>> edge_groups;
+    edge_groups.push(GetNextPropagationEdges(graph, edges_after_dq[0]));
+
+    while (!edge_groups.empty()) {
+      const InlinedVector<ExtendedGraphEdge> curr_edge_group = std::move(edge_groups.front());
+      edge_groups.pop();
+
+      if (curr_edge_group.empty() || any_edge_ends_in_q(graph, curr_edge_group)) {
```

Comment:
if there's an edge that goes to a Q, why do we skip all of `curr_edge_group` instead of just the edge(s) going to a Q?

Reply:
I'm currently skipping scenarios in which one of the edges goes to a Q node for simplicity (and consistency with previous behavior with one edge). It requires more specialized handling, and we don't necessarily need it right now. I added a new comment to note that we skip it.

To go into more detail, here are the implementation options for the following example scenario:
```
DQ -> Transpose --+--> QuantLinear_0 ->
                  |
                  +--> OpA
                  |
                  +--> OpB
```
**Option 1**
Insert `Q -> DQ` ops and ignore the fact that one of the edges goes to an existing Q node:
```
DQ -> Transpose -> Q_new--+--> DQ_new0 -> QuantLinear_0 ->
                          |
                          +--> DQ_new1 -> OpA
                          |
                          +--> DQ_new2 -> OpB
```
This is technically valid but adds unnecessary ops before QuantLinear_0. Also, it is inconsistent with the current behavior in main, which does not propagate a DQ for the sequence `DQ -> Transpose -> QuantLinear_0`.

**Option 2**
Insert `Q -> DQ` only for edges that do not go to the preexisting QuantizeLinear_0:
```
DQ -> Transpose --+--> QuantLinear_0 ->
                  |
                  +--> Q_new --+--> DQ_new0 -> OpA
                               |
                               +--> DQ_new1 -> OpB
```

This _looks_ fine but would not result in a valid QDQ node unit for the Transpose. Our QDQ node grouping rules require the Transpose to have one consumer. So, this option wouldn't be practical.

**Option 3**
If the preexisting QuantizeLinear_0 has scale/zp that match the initial DQ, then we could move OpA and OpB to be downstream from QuantizeLinear_0:
```
DQ -> Transpose -> QuantLinear_0 --+-->
                                   |
                                   +--> DQ_new1 -> OpA
                                   |
                                   +--> DQ_new2 -> OpB
```

I think this option is the best, but it would require more intricate updates. I would prefer to implement this when needed.